InterfaceParser/parsers/rpc_base.py

```diff
@@ -0,0 +1,782 @@
+
+    @abstractmethod
+    def _check_enum_name(self, enum):
+        pass
+
+    def _check_function_param_name(self, function_param_name):
+        """Check function param name.
+
+        This method is called to check whether the newly parsed function
+        parameter name conflicts with some predefined name.
+
+        This implementation doesn't check anything because there is no
+        predefined names in base RPC XML.
+
+        :param function_param_name: string with function name
+        """
+
+    def _parse_root(self, root):
+        """Parse root XML element.
+
+        Default implementation parses root as interface element without a prefix.
+
+        :param root: ElementTree root Element object
+        """
+
+        self._parse_interface(root, "")
+
+    def _parse_interface(self, interface, prefix):
+        """Parse interface element.
+
+        :param interface: interface element.
+        :param prefix: string prefix for all types of the interface.
+        """
+        if interface.tag != "interface":
+            raise ParseError("Invalid interface tag: " + interface.tag)
+
+        params, subelements, attrib = self._parse_base_item(interface, "")
+
+        for param in ["description", "design_description", "todos"]:
+            if len(params[param]) != 0:
+                attrib[param] = "\n".join(params[param])
+
+        if len(params['issues']) != 0:
+            attrib["issues"] = "\n".join(i.value for i in params["issues"])
+
+        self._params.update({prefix + k: v for k, v in attrib.items()})
+
+        for element in subelements:
+            if element.tag == "enum":
+                enum = self._parse_enum(element, prefix)
+                self._check_enum_name(enum)
+                self._add_item(self._enums, enum)
+                self._add_type(enum)
+            elif element.tag == "struct":
+                struct = self._parse_struct(element, prefix)
+                self._add_item(self._structs, struct)
+                self._add_type(struct)
+            elif element.tag == "function":
+                func = self._parse_function(element, prefix)
+                self._add_item(self._functions, func,
+                               (func.function_id, func.message_type))
+            else:
+                raise ParseError("Unexpected element: " + element.tag)
+
+    @staticmethod
+    def _add_item(items, item, key=None):
+        """Add new item in the items dictionary with given key.
+
+        Performs additional check for presence in the dictionary and throws
+        ParseError exception if key already exist.
+
+        :param items: OrderedDict with already parsed EnumElement object
+        :param item: EnumElement object
+        :param key: None
+        """
+        if key is None:
+            key = item.name
+
+        if key in items:
+            raise ParseError("{} '{}' is declared more than once".format(type(item).__name__, key))
+        items[key] = item
```

Comment:
For consistency with `_add_type`, do you think it would be better to do  `self._items[key] = item` here instead of passing self._items as a param?

Reply:
I believe that such deep refactoring of core logic are require additional time for investigation and particular testing. As well are you sure that this change proposal right place for such refactoring?

Reply:
`_add_item` renamed to `_evaluate_element`. This function used independently for evaluating enums, structs, functions appropriately, so approach with combined self._items[key] not applicable here